### ISSUE TITLE
Is there a filename variable?

### ISSUE THREAD

**Issue by Davexx** [daddel80/notepadpp-multireplace#61]
Hi,
I have been playing with MultiReplace for a few days, impressed with the power it has.
I have yet to venture in to the realm of variables.

I was wondering, is there a filename variable where I could replace internal file contents with the file's name?

For example,
Filenames:
John Smith - Payroll.xml
Lisa Borat - Human Resources.xml
etc...

**Search** in a document for "personname" **replace** with part of the filename BEFORE the hyphen.
**Search** in a document for "department", **replace** with part of the filename AFTER the hyphen.

Thanks,
Davexx

**Comment by daddel80** [comment]
Hi Davexx,

Thank you for your suggestion! Yes, it’s a good idea to add the filename as a variable. I’ll implement this and provide you with a version in the next few days. I’m glad to hear you’re using and enjoying MultiReplace!

Best
Thomas

**Comment by daddel80** [comment]
Hi Davexx,

There are now two additional variables: **FNAME** and **FPATH**.

- **FNAME**: Displays the filename or window title (if it's a new, unsaved window).
- **FPATH**: Shows the complete path along with the filename. It will be empty if the file is in a new, unsaved window.

## Solution for Your Filename Example

You can use the following lines in your list. There are two different Lua approaches for the first line—both achieve the same result. Copy / Paste the lines into the list. Choose one of them for the first line:

### First Option:
```lua
1,"","init({ VpersonName = FNAME:sub(1, (FNAME:find("" - "", 1, true) or 0) - 1), Vdepartment = FNAME:sub((FNAME:find("" - "", 1, true) or #FNAME + 1) + 3, (FNAME:find(""."", 1, true) or 0) - 1) })",0,0,1,0,0,""
```

### Second Option:
```lua
1,"","init({ VpersonName = FNAME:match(""^(.-)%s*%-%s*""), Vdepartment = FNAME:match(""%-%s*(.-)%.[^%.]+$"")})",0,0,1,0,0,""


```


### Second Line:
```lua
1,"\\b(personName|department)\\b","cond(MATCH == ""personname"", VpersonName, Vdepartment)",0,0,1,0,1,"\r"
```

1. Switch on **"Replace All in Docs"**.
2. If there is a filename 'John Smith - Payroll.xml' This will replace all occurrences of:
   - `'personname'` with 'John Smith'.
   - `'department'` with 'Payroll'.
   in this file only.

You can find the [Binary here](https://github.com/daddel80/notepadpp-multireplace/actions/runs/12100253175). Just unzip it and copy it into the folder:

`C:\Program Files\Notepad++\plugins\MultiReplace`

**Enjoy!**

**Comment by Davexx** [comment]
I was so thrilled to see your update!
I am having trouble getting it to work, though.
Updated the MultiReplace.dll in c:\Program Files\Notepad++\plugins\MultiReplace\ with the MultiReplace-Release-x64.zip. 
I used your examples above exactly as they are.  
I made sure that "Use Variables" and "Regular Expressions" is turned on...
Attached is a screen shot:
![Screenshot 2024-11-30](https://github.com/user-attachments/assets/e62739d4-c777-4140-b4e0-05121544f3dc)
I noted that the initial " (quotation) was stripped off the front and back of the examples after reload.  Is this okay?
Any suggestions would be very helpful.  I am not adept at LUA and these variables.  I am a novice level at regular expressions.

Gratefully,
Davexx

**Comment by daddel80** [comment]
**Hi, Davexx,**

I already see the error. The copy/paste advice is really vague as a description. Sorry for that. I meant the following. As a result, there should be two lines in the list. For that:

1. Copy the complete line **Option 1** or **Option 2**.
2. Right-click in the list, and you should see **"Paste"** enabled when openeing the Contxmenu by Right-Clicking into the list. This is because MultiReplace recognizes it as a known format. Just paste it, and all settings will also be added automatically. 
![image](https://github.com/user-attachments/assets/5ec67393-bc85-41ba-b2af-220e498d7748)

3. Do the same with the second line.

**How it works:**  
The order of the columns should be irrelevant, as the line with the empty search field will be triggered once when the replacement starts. This prepares the variables for the second-line replacement, which will jump from finding to finding for the replacement.

**But just in case it still seems not to work:**

1. Check the **About** window, which should show **Version 4.1.1.21**.
2. Upload the attached file using the **Load List** button:  
   [Replace_personname_department.csv](https://github.com/user-attachments/files/17968480/Replace_personname_department.csv)
3. Your result should look like this:  
   ![image](https://github.com/user-attachments/assets/7bc1b69d-3ef1-4a4c-bbad-8c701c8a1cb6)
4. Check it by using **"Replace all"** or **"Replace all in all opened documents."**

If it is still not the expected result, just tell me as there are much more things we can dive into!

**Comment by Davexx** [comment]
Thomas,
That worked perfectly!.  Thank you for your patience and guidance on this one, and especially for adding the filename variable!  This is quite a powerful feature, I feel.

I stumbled upon your tool when using find & replace in Notepad++.  It was getting very complicated and I was wishing there was a more robust way to do find & replace, where I could have multiple instances, be able to edit them in a list, be able to turn them on and off.  Out of exhaustion, I went to the plugins database and searched for "replace" and that's where your full-featured plug-in was introduced to me. 
I really wish this was part of Notepad++!

One last question:
If I wanted to make the detection of the two elements in the filename, currently separated by the "-" (hyphen) a bit more robust for cases where the naming was not as precise, what would you suggest for the first replace line?
For example, if the files looked like this:
John Smith - Payroll.xml
Barry Henkel-Engineering.xml
Lisa Borat- Human Resources.xml
Angela Stoltz -Maintenance.xml
(Note that the hyphen does not have consistent spacing around it.)

Again, thank you for your excellent plugin and support!
Davexx

**Comment by daddel80** [comment]
Hi Davexx,

good to hear that it worked fine for you, and thank you for your kind words. I really appreciate it! I'm glad to hear that the plugin has been helpful for your use case.

To make the detection of the two elements in the filename more robust, even with inconsistent spacing around the hyphen, you can use the following line:

```lua
1,"","init({ VpersonName = FNAME:match(""^(.-)%s*%-+%s*(.*)$""), Vdepartment = FNAME:match(""^.-%s*%-+%s*(.*)%..-$"") })",0,0,1,0,0,""
```

If you don’t mind, I’d like to ask you for a small favor. I noticed in your previous screenshot that the extended phrase wasn’t fully visible. Could you install a corrected version you can [download here](https://github.com/daddel80/notepadpp-multireplace/actions/runs/12108237826) and please share a new screenshot of MultiReplace Window? Scaling with font sizes in Windows is always a bit tricky! Thxs

Cheers,  
Thomas

**Comment by daddel80** [comment]
Hi Davexx,
Yes, this is what I was looking for, but I see that the space I added for "Extended" is still not enough—the closing bracket is still not visible. Unfortunately, I have to estimate the required space as I cannot reproduce the issue in my environment.

I will provide you with a new version tomorrow and hope this adjustment will finally make everything fit properly.

**Comment by Davexx** [comment]
Thomas,
I see what you mean.  "Extended" has its example variable cut off.
Yes, Windows has tricky display scaling and font scaling.
I am running a 4K monitor set to 150% scaling as my primary and a 2560 x 1440 set to 100% scaling as my secondary.  I took the screenshot on the secondary monitor, with MultiReplace.ini `ScaleFactor=1.0`. Normally I run it `ScaleFactor=1.5` as it is a challenge for me to see.  But line length vs. font size is always a challenge.
May I suggest a couple of GUI refinements possibilities?

- Move the "Load/Save/Export List" and "Move Lines" buttons to the right of the "Replace/Find/Mark" items so the list section is as wide as it can be.
- Is it possible to "word wrap" the Find and replace items?  If you think that might confuse users, then:
- Add a tooltip mouse hover to the find and replace lines so that their contents can always be seen, without adjusting the column widths or without using "Edit Field".

Question: Is there a way to toggle the WCVER items after they are committed to the lower list?  When I first started using MultiReplace I was wishing there was a way to do that.

**Comment by daddel80** [comment]
Hi Davexx,  
thanks for your input, I really appreciate it.  

- Thanks for sharing the scaling information, i would be curious t o know. Therefore, there’s no need to check this via **F12**. However, if you press **F12** on any selected line in the list, a window will pop up showing all the details.
- I shifted everything a bit more to the right again. Let’s see if the "Extended" phrase fits properly now.  
- I think "word wrapping" in the **Find** and **Replace** fields would probably confuse most users.  
- The tooltip idea on mouse hover is way better. I’m planning to implement it. 

Answer: Yes, just double-click on any field in the list, and you can change its value.  

I’d also like to provide you with a new binary. This will be **release 4.2**, so one step further. It already includes Undo and Redo for list entries. It’s not fully tested yet, but it should run stable. All the new variables (**FNAME** and **FPATH**) are included. Please check it especially for the **Extended** field, and send me a screenshot if possible.  You can[ download it here.](https://github.com/daddel80/notepadpp-multireplace/actions/runs/12108204492)

**Comment by Davexx** [comment]
Thomas,
I am embarrassed that I did not discover the WCVER toggle by  double-clicking.  I only single clicked...  Hopefully that can make it to the documentation for people like me! :-S

I did a super screenshot with all the info in it (below).  I set my **MultiReplace.ini** `ScaleFactor=1.0`.  Note that the **F12** does not tell the exact truth.  As stated before, my primary (1) monitor is 3840 x 2160 with a scale of 150%, and my secondary (2) monitor is 2560 x 1440 Scale at 100%, so the **F12** seems to not properly report the secondary statistics.

![Screenshot 2024-12-02](https://github.com/user-attachments/assets/f3bbdb0a-e596-4729-b922-e8e270d6b866)

Thanks for considering the tooltip mouse-over idea.  I think that will be very helpful.

Dave.
